src/v4l/mod.rs

```diff
@@ -0,0 +1,249 @@
+///
+/// # Neolink V4l
+///
+/// This module serves the v4l streams for the
+/// `neolink v4l` subcommand
+///
+/// All camera specified in the config.toml will be served
+/// over v4l.
+///
+/// You can view the streams with any v4l compliement program
+/// such as ffmpeg, vlc, blue-iris, home-assistant, zone-minder etc.
+///
+/// `/dev/video0`
+///
+/// # Usage
+///
+/// To start the subcommand use the following in a shell.
+///
+/// ```bash
+/// neolink v4l --config=config.toml
+/// ```
+///
+use anyhow::{Context, Result};
+use log::*;
+use neolink_core::bc_protocol::{BcCamera, Stream};
+use neolink_core::Never;
+use std::sync::Arc;
+use std::time::Duration;
+
+// mod adpcm;
+/// The command line parameters for this subcommand
+mod cmdline;
+/// The errors this subcommand can raise
+mod v4lt;
+
+use super::config::{CameraConfig, Config};
+use crate::utils::AddressOrUid;
+pub(crate) use cmdline::Opt;
+use v4lt::{V4ltOutputs, V4lDevice};
+
+/// Entry point for the v4l subcommand
+///
+/// Opt is the command line options
+pub(crate) fn main(_opt: Opt, config: Config) -> Result<()> {
+    if config.certificate == None && !config.users.is_empty() {
+        warn!(
+            "Without a server certificate, usernames and passwords will be exchanged in plaintext!"
+        )
+    }
+
+    crossbeam::scope(|s| {
+        for camera in config.cameras {
+            if camera.format.is_some() {
+                warn!("The format config option of the camera has been removed in favour of auto detection.")
+            }
+            // Let subthreads share the camera object; in principle I think they could share
+            // the object as it sits in the config.cameras block, but I have not figured out the
+            // syntax for that.
+            let arc_cam = Arc::new(camera);
+
+            if ["mainStream"].iter().any(|&e| e == arc_cam.v4lstream) {
+                let v4l = V4lDevice::new(arc_cam.v4ldevice as usize);
+                let mut outputs = v4l
+                    .add_stream()
+                    .unwrap();
+                let main_camera = arc_cam.clone();
+                s.spawn(move |_| camera_loop(&*main_camera, Stream::Main, &mut outputs, true));
```

Comment:
When you make `arc_cam.v4ldevice` an `Option<u8>` as suggested above you'd need to make this conditional like this:

```rust
if let Some(v4ldevice) = arc_cam.v4ldevice {
  let v4l = V4lDevice::new(v4ldevice as usize);
  let mut outputs = v4l
      .add_stream()
      .unwrap();
  let main_camera = arc_cam.clone();
  s.spawn(move |_| camera_loop(&*main_camera, Stream::Main, &mut outputs, true));
}
```

Reply:
For the other ones too, of just one big if outside all of them